File: tools/lib/iomisc/reader_test.go
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package iomisc

import (
	"io"
	"strings"
	"testing"
)

func TestReaderAtToReader(t *testing.T) {
	r := ReaderAtToReader(strings.NewReader("123456789"))
	if _, ok := r.(io.ReaderAt); ok {
		t.Fatalf("reader implements io.ReaderAt when it shouldn't")
	}
	tests := []struct {
		byteLen        int
		expectedErr    error
		expectedOutput string
	}{
		{byteLen: 4, expectedErr: nil, expectedOutput: "1234"},
		{byteLen: 4, expectedErr: nil, expectedOutput: "5678"},
		{byteLen: 4, expectedErr: io.EOF, expectedOutput: "9"},
		{byteLen: 4, expectedErr: io.EOF, expectedOutput: ""},
	}
	for _, test := range tests {
		buf := make([]byte, test.byteLen)
		n, err := r.Read(buf)
		if err != test.expectedErr {
			t.Errorf("got err: %v, expected: %v", err, test.expectedErr)
		}
		if n != len(test.expectedOutput) {
			t.Errorf("read %d bytes, expected: %d", n, len(test.expectedOutput))
		}
		if string(buf[:n]) != test.expectedOutput {
			t.Errorf("read: %s, expected: %s", string(buf), test.expectedOutput)
		}
	}
}

File: tools/symbol-index/reader_test.cc
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "tools/symbol-index/reader.h"

#include <sstream>

#include <gtest/gtest.h>

#include "src/lib/fxl/strings/split_string.h"

namespace symbol_index {

namespace {

const char* test_content = R"(
# This is a comment and should be ignored
# Empty lines should also be ignored
   
abc 

# This is an intermediate comment that should also be ignored.
   abcd   efgh
)";

TEST(ReaderTest, Read) {
  Reader reader(' ');
  std::vector<std::vector<std::string>> output;
  std::stringstream file(test_content);

  ASSERT_TRUE(reader.Read(file, "", &output).empty());
  ASSERT_EQ(output.size(), 2UL);
  ASSERT_EQ(output[0].size(), 1UL);
  ASSERT_EQ(output[0][0], "abc");
  ASSERT_EQ(output[1].size(), 2UL);
  ASSERT_EQ(output[1][0], "abcd");
  ASSERT_EQ(output[1][1], "efgh");
}

}  // namespace

}  // namespace symbol_index
